src/Equinox.Cosmos/Cosmos.fs

```diff
@@ -277,53 +292,157 @@ module Log =
     let (|BlobLen|) = function null -> 0 | (x : byte[]) -> x.Length
+    let (|EventLen|) (x: #Store.IEvent) = let (BlobLen bytes), (BlobLen metaBytes) = x.Data, x.Meta in bytes+metaBytes
+    let (|BatchLen|) = Seq.sumBy (|EventLen|)
 
 [<RequireQualifiedAccess; NoEquality; NoComparison>]
 type EqxSyncResult =
     | Written of Store.Position
+    | Conflict of Store.Position * events: Store.IOrderedEvent[]
     | ConflictUnknown of Store.Position
-    | Conflict of Store.Position * events: Store.IEventData[]
 
 // NB don't nest in a private module, or serialization will fail miserably ;)
 [<CLIMutable; NoEquality; NoComparison; JsonObject(ItemRequired=Required.AllowNull)>]
-type WriteResponse = { etag: string; conflicts: Store.IndexProjection[] }
+type WriteResponse = { etag: string; i: int64; conflicts: Store.Projection[] }
 
 module private Write =
     let [<Literal>] sprocName = "EquinoxIndexedWrite"
+    let [<Literal>] sprocBody = """
+    function indexedWrite(batch) {
+        var response = getContext().getResponse();
+        var collection = getContext().getCollection();
+        var collectionLink = collection.getSelfLink();
+
+        if (batch === undefined || batch==null) {
+            response.setBody({ etag: null, i: -2, conflicts: null });
+        }
+        else {
+            tryQueryAndUpdate();
+
+            // Recursively queries for a document by id w/ support for continuation tokens.
+            // Calls tryUpdate(document) as soon as the query returns a document.
+            function tryQueryAndUpdate(continuation) {
+                var query = {query: "select * from root r where r.id = @id and r.p = @p", parameters: [{name: "@id", value: "-1"},{name: "@p", value: batch.p}]};
+                var requestOptions = {continuation: continuation};
+
+                var isAccepted = collection.queryDocuments(collectionLink, query, requestOptions, function (err, documents, responseOptions) {
+                    if (err) throw new Error("Error" + err.message);
+
+                    if (documents.length > 0) {
+                        // If the document is found, update it.
+                        // There is no need to check for a continuation token since we are querying for a single document.
+                        tryUpdate(documents[0], false);
+                    } else if (responseOptions.continuation) {
+                        // Else if the query came back empty, but with a continuation token; repeat the query w/ the token.
+                        // It is highly unlikely for this to happen when performing a query by id; but is included to serve as an example for larger queries.
+                        tryQueryAndUpdate(responseOptions.continuation);
+                    } else {
+                        // Else the snapshot does not exist; create snapshot
+                        var doc = {p:batch.p, id:"-1", i:-1, e:[], c:[]};
```

Comment:
`i` should not ever be -1 - better to not set it

Reply:
The incoming batch is intended to be used directly in this case, replacing with that